File: leet_binary_tree_right_side_view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        # s1 = [root]
        # s2 = []
        # res = []
        # while s1:
        #     res.append(s1[0].val)
        #     s2 = []
        #     for u in s1:
        #         if u.right:
        #             s2.append(u.right)
        #         if u.left:
        #             s2.append(u.left)
        #     s1=s2
        def rec(root,level):
            if not root:
                return
            
            if len(res)<level:
                res.append(root.val)
            rec(root.right,level+1)
            rec(root.left,level+1)
            
        res = []
        rec(root,1)
        return res

File: LearnJava/Leet_Binary_Tree_Right_Side_View.java
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
 
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<>();
        rightSideViewUtil(root, 0, result);
        return result;
    }
    
    public void rightSideViewUtil(TreeNode node, int level, ArrayList<Integer> result) {
        if (node == null) {
            return;
        }
        
        if (level == result.size()) {
            result.add(node.val);
        }
//         else {
//             result.set(level, node.val);
//         }
        
        rightSideViewUtil(node.right, level+1, result);
        rightSideViewUtil(node.left, level+1, result);
        
    }
}